https://www.w3schools.com/python/default.asp

# Creating Variables

Variables do not need to be declared with any particular type, and can even change type after they have been set.

## Case-Sensitive
Variable names are case-sensitive.

In [1]:
x = 5 # x is of type int
y = "John" # x is now of type str
X = 4
print(x)
print(y)
print(X)

5
John
4


## Casting
If you want to specify the data type of a variable, this can be done with casting

In [2]:
x =str(3)
y = int(3)
z= float(3)
type(x) , type(y) , type(z)

(str, int, float)

## Variable Names
A variable can have a short name (like x and y) or a more descriptive name (age, carname, total_volume). 
###  Rules for Python variables:
1. A variable name must start with a letter or the underscore character
2. A variable name cannot start with a number
3. A variable name can only contain alpha-numeric characters and underscores (A-z, 0-9, and _ )
4. Variable names are case-sensitive (age, Age and AGE are three different variables)

In [3]:
myvar = "John"
my_var = "John"
_my_var = "John"
myVar = "John"
MYVAR = "John"
myvar2 = "John"

## Multi Words Variable Names
Variable names with more than one word can be difficult to read.

There are several techniques you can use to make them more readable:

### Camel Case
Each word, except the first, starts with a capital letter:

In [4]:
myVariableName = "John"

### Pascal Case
Each word starts with a capital letter:

In [5]:
MyVariableName = "John"

### Snake Case
Each word is separated by an underscore character:

In [6]:
my_variable_name = "John"

## Many Values to Multiple Variables
Python allows you to assign values to multiple variables in one line:

In [7]:
x , y , z = "Oregon" , "Milaukee" , "Delvar"
print(x , y , z)

Oregon Milaukee Delvar


## One Value to Multiple Variables
And you can assign the same value to multiple variables in one line:



In [8]:
x = y = z = "Orange"
print (x,  y  , z)

Orange Orange Orange


## Unpack a Collection
If you have a collection of values in a list, tuple etc. Python allows you extract the values into variables. This is called unpacking.

In [9]:
furits = ["apple", "banana", "cherry"]
x , y , z = furits
print (x,  y  , z)

apple banana cherry


## Output Variables
The Python print statement is often used to output variables.

To combine both text and a variable, Python uses the `+` character:

In [10]:
x = "Awesome"
print("Python  is " + x)

Python  is Awesome


In [11]:
x = "Ali"
y = " Nemati"
z =  x + y
print(z)

Ali Nemati


In [12]:
x = 5
y = 10
print(x + y)

15


## Global Variables
Variables that are created outside of a function (as in all of the examples above) are known as global variables.

Global variables can be used by everyone, both inside of functions and outside.

Create a variable outside of a function, and use it inside the function



In [13]:
x = "Awesome"

def myfunc():
    print("Python is " + x)

myfunc()

Python is Awesome


## Local Variables


If you create a variable with the same name inside a function, this variable will be local, and can only be used inside the function. The global variable with the same name will remain as it was, global and with the original value.



In [14]:
x = "Awesome"

def myfunc():
    x = " Fantastic"
    print("Python is " + x)

myfunc()

Python is  Fantastic


## The global Keyword
Normally, when you create a variable inside a function, that variable is local, and can only be used inside that function.

To create a global variable inside a function, you can use the **global** keyword.

In [15]:
x = "Test"
def func():
    global x
    x = "Fantastic for EVER!"



print("ptyhon is " + x)

ptyhon is Test


In [16]:
x = "Test"
def func():
    global x
    x = "Fantastic for EVER!"

func()

print("ptyhon is " + x)

ptyhon is Fantastic for EVER!


# Python Data Types
Built-in Data Types
In programming, data type is an important concept.

Variables can store data of different types, and different types can do different things.

Python has the following data types built-in by default, in these categories:
    
| data type | Description |
|-----------|-------------|
| Text Type: | str |
| Numeric Types: | int, float, complex |
| Sequence Types: | list, tuple, range |
| Mapping Type: | dict |
| Set Types: | set, frozenset |
| Boolean Type: | bool |
| Binary Types: | bytes, bytearray, memoryview |

## Python Numbers
There are three numeric types in Python:

### 1. int : Int, or integer, 

is a whole number, positive or negative, without decimals, of unlimited length.

### 2. float : 

Float, or "floating point number" is a number, positive or negative, containing one or more decimals. Float can also be scientific numbers with an "e" to indicate the power of 10.

### 3. complex : 
Complex numbers are written with a "j" as the imaginary part:

### Underscores in Numbers

When you’re writing long numbers, you can group digits using underscores
to make large numbers more readable.

universe_age = 14_000_000_000

In [17]:
0.1 + 0.1

0.2

In [18]:
4/2

2.0

In [19]:
x = 5
y = 2.8 
z =2j
w =  -87.7e100

print(x , " , " ,  type(x))
print(y , " , " ,  type(y))
print(z , " , " ,  type(z))
print(w , " , " ,  type(w))

5  ,  <class 'int'>
2.8  ,  <class 'float'>
2j  ,  <class 'complex'>
-8.77e+101  ,  <class 'float'>


## Type Conversion
You can convert from one type to another with the int(), float(), and complex() methods:

In [20]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

#convert from int to float:
a = float(x)
print(a , type(a))
#convert from float to int:
b = int(y)
print(b , type(b))
#convert from int to complex:
c = complex(x )
print(c ,  type(c))


1.0 <class 'float'>
2 <class 'int'>
(1+0j) <class 'complex'>


## Random Number
Python does not have a random() function to make a random number, but Python has a built-in module called random that can be used to make random numbers:



In [21]:
import random
print(random.randrange(1, 200))

76


## Casting

There may be times when you want to specify a type on to a variable. This can be done with casting. Python is an object-orientated language, and as such it uses classes to define data types, including its primitive types.

Casting in python is therefore done using constructor functions:

* int() - constructs an integer number from an integer literal, a float literal (by removing all decimals), or a string literal (providing the string represents a whole number)
* float() - constructs a float number from an integer literal, a float literal or a string literal (providing the string represents a float or an integer)
* str() - constructs a string from a wide variety of data types, including strings, integer literals and float literals

In [22]:
x = int(1)   # x will be 1
y = int(2.8) # y will be 2
z = int("3") # z will be 3
print(x, y , z)
print(type(x), type(y) , type(z))

1 2 3
<class 'int'> <class 'int'> <class 'int'>


In [23]:
x = float(1)     # x will be 1.0
y = float(2.8)   # y will be 2.8
z = float("3")   # z will be 3.0
w = float("4.2") # w will be 4.2
print(x, y , z , w)
print(type(x), type(y) , type(z) , type(w))

1.0 2.8 3.0 4.2
<class 'float'> <class 'float'> <class 'float'> <class 'float'>


In [24]:
x = str("s1") # x will be 's1'
y = str(2)    # y will be '2'
z = str(3.0)  # z will be '3.0'
print(x, y , z)
print(type(x), type(y) , type(z))

s1 2 3.0
<class 'str'> <class 'str'> <class 'str'>


## Strings
Strings in python are surrounded by either single quotation marks, or double quotation marks.

`'hello'` is the same as `"hello"`.

You can display a string literal with the print() function:

In [25]:
name = "ada lovelace"
print(name.title())

x = str("Hello World")
print(x , " , " ,  type(x))

Ada Lovelace
Hello World  ,  <class 'str'>


## Multiline Strings
You can assign a multiline string to a variable by using three quotes:

In [26]:
a = """Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua."""
print(a)

Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.


In [27]:
a

'Lorem ipsum dolor sit amet,\nconsectetur adipiscing elit,\nsed do eiusmod tempor incididunt\nut labore et dolore magna aliqua.'

# Strings are Arrays
strings in Python are arrays of bytes representing unicode characters.

However, Python does not have a character data type, a single character is simply a string with a length of 1.

Square brackets can be used to access elements of the string.

In [28]:
a[1] , a[20:30] ,  a[70:] ,  a[:-70]

('o',
 't amet,\nco',
 'd tempor incididunt\nut labore et dolore magna aliqua.',
 'Lorem ipsum dolor sit amet,\nconsectetur adipiscing el')

## Looping Through a String
Since strings are arrays, we can loop through the characters in a string, with a for loop.

In [29]:
for x in "banana":
    print(x)

b
a
n
a
n
a


## String Length
To get the length of a string, use the len() function.

In [30]:
len(a)

123

## Check String
To check if a certain phrase or character is present in a string, we can use the keyword in.

In [31]:
print("free" in a)
print("incididunt" in a)
print("expensive" not in a)
if "incididunt" in a:
    print("Yes, we can find it.")
else:
    print("Nope!")

False
True
True
Yes, we can find it.


## Slicing
You can return a range of characters by using the slice syntax.

Specify the start index and the end index, separated by a colon, to return a part of the string.

In [32]:
b = "Hello, World!"
print(b[2:8])
print(b[:5]) #Slice From the Start
print(b[2:]) # Slice To the End
print(b[-5:-2]) # Negative Indexing

llo, W
Hello
llo, World!
orl


## Modify Strings
Python has a set of built-in methods that you can use on strings.
### Upper Case
### Lower Case
### Remove Whitespace
The strip() method removes any whitespace from the beginning or the end.

In [33]:
a = " Hello, World!"
print(a.upper() , a.lower() , a.strip())
name = "Ada Lovelace"
print(name.upper())
print(name.lower())

 HELLO, WORLD!  hello, world! Hello, World!
ADA LOVELACE
ada lovelace


### Replace String

In [34]:
a = " Hello, World!"
a.replace("l", "M")

' HeMMo, WorMd!'

## Split String
The split() method returns a list where the text between the specified separator becomes the list items.

In [35]:
a.split("o")

[' Hell', ', W', 'rld!']

## String Concatenation
To concatenate, or combine, two strings you can use the + operator.

In [36]:
a = "Ali"
b = "World"
c = a + " " + b
print(c)

Ali World


## String Format
The format() method takes the passed arguments, formats them, and places them in the string where the placeholders {} are:

In [37]:
age= 35
txt = "My name is Ali and I am {} years old"
print(txt.format(age))

My name is Ali and I am 35 years old


In [38]:
first_name = "ada"
last_name = "lovelace"
full_name = f"{first_name} {last_name}"
print(f"Hello, {full_name.title()}!")

Hello, Ada Lovelace!


In [39]:
first_name = "ada"
last_name = "lovelace"
full_name = f"{first_name} {last_name}"
message = f"Hello, {full_name.title()}!"
print(message)

Hello, Ada Lovelace!


In [40]:
quantity = 3
itemno = 567
price = 49.95
myorder = "I want {} pieces of item {} for {} dollars."
print(myorder.format(quantity, itemno, price))

I want 3 pieces of item 567 for 49.95 dollars.


You can use index numbers {0} to be sure the arguments are placed in the correct placeholders:

In [41]:
quantity = 4
itemno = 5567
price = 494.95
myorder = "I want to pay {2} dollars for {0} pieces of item {1}."
print(myorder.format(quantity, itemno, price))

I want to pay 494.95 dollars for 4 pieces of item 5567.


## Escape Characters
| Code | Result |
|------|--------| 
| \\ | Backslash |
| \n | New Line |
| \r | Carriage Return  |
| \t | Tab |
| \b | Backspace |
| \f | Form Feed |
| \ooo | Octal value  |
| \xhh | Hex value  |

In [42]:
print("Languages:\nPython\nC\nJavaScript")

Languages:
Python
C
JavaScript


In [43]:
## Zen of Python: Enter import this into a Python terminal session and skim
# through the additional principles.

import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


In [44]:
print("\tPython")

	Python


# Constants

A constant is like a variable whose value stays the same throughout the life
of a program. Python doesn’t have built-in constant types, but Python programmers use all capital letters to indicate a variable should be treated as a
constant and never be changed:

MAX_CONNECTIONS = 5000

## Booleans
Booleans represent one of two values: True or False.

In programming you often need to know if an expression is True or False.

You can evaluate any expression in Python, and get one of two answers, True or False.

When you compare two values, the expression is evaluated and Python returns the Boolean answer:

In [45]:
x = bool(5)
print(x , " , " ,  type(x))
print(bool("Hello"))
print(bool(15))

True  ,  <class 'bool'>
True
True


In [46]:
print(10 > 9)
print(10 == 9)
print(10 < 9)

True
False
False


In [47]:
a = 500
b = 33
if b > a:
    print("B is greater than a")
else:
    print("B is not greater than a")

B is not greater than a


In [48]:
x = "Hello"
y = 15

print(bool(x))
print(bool(y))

True
True


### Most Values are True
Almost any value is evaluated to True if it has some sort of content.

Any string is True, except empty strings.

Any number is True, except 0.

Any list, tuple, set, and dictionary are True, except empty ones.

In [49]:
print(bool("abc"))
print(bool(123))
print(bool(["apple", "cherry", "banana"]))
print("------------------")
print(bool(False))
print(bool(None))
print(bool(0))
print(bool(""))
print(bool(()))
print(bool([]))
print(bool({}))

True
True
True
------------------
False
False
False
False
False
False
False


In [50]:
def myFunction() :
  return True

print(myFunction())

True


In [51]:
def myFunction() :
  return True

if myFunction():
  print("YES!")
else:
  print("NO!")

YES!


In [52]:
x = 200
print(isinstance(x, int))

True


# Operators
Operators are used to perform operations on variables and values.

In [53]:
print(10 + 5)

15


Python divides the operators in the following groups:

### Arithmetic operators

![](./i/Arithmeticoperators.jpg)

### Assignment operators

![](./i/Assignment_Operators.jpg)

### Comparison operators

![](./i/Comparison_Operators.jpg)

### Logical operators

![](./i/Logical_Operators.jpg)

### Identity operators

![](./i/Identity_Operators.jpg)

### Membership operators


![](./i/Membership_operators.jpg)

###  Bitwise operators

![](./i/Bitwise_operators.jpg)


# Lists

List items are ordered, changeable, and allow duplicate values.

List items are indexed, the first item has index [0], the second item has index [1] etc.

## Ordered

When we say that lists are ordered, it means that the items have a defined order, and that order will not change.

If you add new items to a list, the new items will be placed at the end of the list.

## Changeable

The list is changeable, meaning that we can change, add, and remove items in a list after it has been created.

## Allow Duplicates

Since lists are indexed, lists can have items with the same value:

## List Length

To determine how many items a list has, use the len() function

## Python has a set of built-in methods that you can use on lists.

| Method | Description |
|--------|-------------| 
| append() | Adds an element at the end of the list | 
| clear() | Removes all the elements from the list | 
| copy() | Returns a copy of the list | 
| count() | Returns the number of elements with the specified value | 
| extend() | Add the elements of a list (or any iterable), to the end of the current list | 
| index() | Returns the index of the first element with the specified value | 
| insert() | Adds an element at the specified position | 
| pop() | Removes the element at the specified position | 
| remove() | Removes the item with the specified value | 
| reverse() | Reverses the order of the list | 
| sort() | Sorts the list | 

In [54]:
x = list(("apple", "banana", "cherry"))
print(x , x[1] , " , " ,  type(x))

['apple', 'banana', 'cherry'] banana  ,  <class 'list'>


In [55]:
thislist = ["apple", "banana", "cherry", "apple", "cherry"]
print(thislist)

['apple', 'banana', 'cherry', 'apple', 'cherry']


In [56]:
thislist = ["apple", "banana", "cherry"]
print(len(thislist))

3


## List Items - Data Types
List items can be of any data type:

In [57]:
list1 = ["apple", "banana", "cherry"]
list2 = [1, 5, 7, 9, 3]
list3 = [True, False, False]
print(list1)
print(list2)
print(list3)


['apple', 'banana', 'cherry']
[1, 5, 7, 9, 3]
[True, False, False]


In [58]:
list1 = ["abc", 34, True, 40, "male"]
type(list1)

list

## The list() Constructor
It is also possible to use the list() constructor when creating a new list.

In [59]:
thislist = list(("apple", "banana", "cherry")) # note the double round-brackets
print(thislist)

['apple', 'banana', 'cherry']


## There are four collection data types in the Python programming language:

* List is a collection which is ordered and changeable. Allows duplicate members.
* Tuple is a collection which is ordered and unchangeable. Allows duplicate members.
* Set is a collection which is unordered and unindexed. No duplicate members.
* Dictionary is a collection which is ordered and changeable. No duplicate members.

## Access Items
List items are indexed and you can access them by referring to the index number.


In [60]:
thislist = ["apple", "banana", "cherry"]
print(thislist[1])

banana


## Negative Indexing
Negative indexing means start from the end.
-1 refers to the last item, -2 refers to the second last item etc.


In [61]:
thislist = ["apple", "banana", "cherry"]
print(thislist[-1])
print(thislist[-3])

cherry
apple


## Range of Indexes

You can specify a range of indexes by specifying where to start and where to end the range.

When specifying a range, the return value will be a new list with the specified items.


Note: The search will start at index 2 (included) and end at index 5 (not included).

Remember that the first item has index 0.



In [62]:
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
print(thislist[2:5])

['cherry', 'orange', 'kiwi']


By leaving out the start value, the range will start at the first item.

In [63]:
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
print(thislist[:4])

['apple', 'banana', 'cherry', 'orange']


By leaving out the end value, the range will go on to the end of the list.

In [64]:
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
print(thislist[2:])

['cherry', 'orange', 'kiwi', 'melon', 'mango']


## Range of Negative Indexes
Specify negative indexes if you want to start the search from the end of the list:

Example

In [65]:
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
print(thislist[-5:-2])

['cherry', 'orange', 'kiwi']


## Check if Item Exists
To determine if a specified item is present in a list use the in keyword:



In [66]:
if "apple" in thislist:
    print("YES")
else:
    print("Nope")

YES


## Change List Items
To change the value of a specific item, refer to the index number:

If you insert more items than you replace, the new items will be inserted where you specified,  and the remaining items will move accordingly:

Note: The length of the list will change when the number of items inserted does not match the number of items replaced.


In [67]:
thislist = ["apple", "banana", "cherry"]
thislist[1] = "Test"
print(thislist)

['apple', 'Test', 'cherry']


In [68]:
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "mango"]
print(thislist)
thislist[1:3] = ["blackcurrant", "watermelon"]
print(thislist)

['apple', 'banana', 'cherry', 'orange', 'kiwi', 'mango']
['apple', 'blackcurrant', 'watermelon', 'orange', 'kiwi', 'mango']


In [69]:
thislist = ["apple", "banana", "cherry"]
thislist[1:2] = ["blackcurrant", "watermelon"]
print(thislist)

['apple', 'blackcurrant', 'watermelon', 'cherry']


## Insert Items
To insert a new list item, without replacing any of the existing values, we can use the insert() method.

The insert() method inserts an item at the specified index:

In [70]:
thislist = ["apple", "banana", "cherry"]
thislist.insert(2, "watermelon")
print(thislist)

# Note: As a result of the example above, the list will now contain 4 items.

['apple', 'banana', 'watermelon', 'cherry']


In [71]:
"watermelon" in thislist

True

## Find Index of item in List

In [72]:
thislist.index("watermelon") , len(thislist)

(2, 4)

## Add List Items
To add an item to the end of the list, use the append() method:

In [73]:
thislist = ["apple", "banana", "cherry"]
thislist.append("orange")
print(thislist)

['apple', 'banana', 'cherry', 'orange']


## Extend List
To append elements from another list to the current list, use the extend() method

In [74]:
thislist = ["apple", "banana", "cherry"]
tropical = ["mango", "pineapple", "papaya" , 1 , 15 , "a1j"]
thislist.extend(tropical)
print(thislist)

['apple', 'banana', 'cherry', 'mango', 'pineapple', 'papaya', 1, 15, 'a1j']


## Remove List Items
The remove() method removes the specified item.

In [75]:
thislist = ["apple", "banana", "cherry"]
thislist.remove("banana")
print(thislist)

['apple', 'cherry']


## Remove Specified Index
The pop() method removes the specified index.
or Del

In [76]:
thislist = ["apple", "banana", "cherry"]
thislist.pop(2)
print(thislist)

['apple', 'banana']


In [77]:
thislist = ["apple", "banana", "cherry"]
del thislist[2]
print(thislist)

['apple', 'banana']


In [78]:
# del Entire list

thislist = ["apple", "banana", "cherry"]
del thislist
# print(thislist) NameError: name 'thislist' is not defined

## Clear the List
The clear() method empties the list.

The list still remains, but it has no content.

In [79]:
thislist = ["apple", "banana", "cherry"]
thislist.clear()
print(thislist)

[]


## Loop Through a List
You can loop through the list items by using a for loop:

In [80]:
thislist = ["apple", "banana", "cherry"]
for i in thislist:
    print(i)

apple
banana
cherry


## Loop Through the Index Numbers
You can also loop through the list items by referring to their index number.

Use the range() and len() functions to create a suitable iterable.

In [81]:
for i in range(len(thislist)):
    print("index " , i , " has " , thislist[i] , " .")

index  0  has  apple  .
index  1  has  banana  .
index  2  has  cherry  .


## Using a While Loop
You can loop through the list items by using a while loop.

Use the len() function to determine the length of the list, then start at 0 and loop your way through the list items by refering to their indexes.

Remember to increase the index by 1 after each iteration.

In [82]:
i = 0
while i < len(thislist):
    print(thislist[i])
    i +=1

apple
banana
cherry


## List Comprehension
List comprehension offers a shorter syntax when you want to create a new list based on the values of an existing list.

```python
newlist = [expression for item in iterable if condition == True]
```
**The return value is a new list, leaving the old list unchanged.**


Example:

Based on a list of fruits, you want a new list, containing only the fruits with the letter "a" in the name.

Without list comprehension you will have to write a for statement with a conditional test inside:

In [83]:
fruits = ["apple", "banana", "cherry", "kiwi", "mango"]
newlist = []

for x in fruits:
    if 'a' in x:
        newlist.append(x)
        
    
print(newlist)



['apple', 'banana', 'mango']


In [84]:
newlist = [x for x in fruits if "a" in x]

print(newlist)

['apple', 'banana', 'mango']


### Condition
The condition is like a filter that only accepts the items that valuate to True.

In [85]:
[x for x in fruits if x != "apple"] # Only accept items that are not "apple"

['banana', 'cherry', 'kiwi', 'mango']

In [86]:
[x for x in fruits] # With no if statement

['apple', 'banana', 'cherry', 'kiwi', 'mango']

### The iterable 
can be any iterable object, like a list, tuple, set etc.

In [87]:
[x for x in range(10)] # 

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [88]:
# Accept only numbers lower than 5:

[x for x in range(10) if x < 5]

[0, 1, 2, 3, 4]

### Expression
The expression is the current item in the iteration, but it is also the outcome, which you can manipulate before it ends up like a list item in the new list:

In [89]:
[x.upper() for x in fruits]

['APPLE', 'BANANA', 'CHERRY', 'KIWI', 'MANGO']

In [90]:
 ['hello' for x in fruits]

['hello', 'hello', 'hello', 'hello', 'hello']

The expression can also contain conditions, not like a filter, but as a way to manipulate the outcome:



In [91]:
[x if x != "banana" else "orange" for x in fruits]
# "Return the item if it is not banana, if it is banana return orange".

['apple', 'orange', 'cherry', 'kiwi', 'mango']

## Sort Lists
List objects have a sort() method that will sort the list alphanumerically, ascending, by default:

In [92]:
thislist = ["orange", "mango", "kiwi", "pineapple", "banana"]
thislist.sort()
print(thislist)
thislist = [100, 50, 65, 82, 23]
thislist.sort()
print(thislist)

['banana', 'kiwi', 'mango', 'orange', 'pineapple']
[23, 50, 65, 82, 100]


In [93]:
thislist = ["orange", "mango", "kiwi", "pineapple", "banana"]
thislist.sort(reverse=True)
print(thislist)
thislist = [100, 50, 65, 82, 23]
thislist.sort(reverse=True)
print(thislist)

['pineapple', 'orange', 'mango', 'kiwi', 'banana']
[100, 82, 65, 50, 23]


## Customize Sort Function
You can also customize your own function by using the keyword argument key = function.

The function will return a number that will be used to sort the list (the lowest number first):

In [94]:
def myfunc(n):
  return abs(n - 50)

thislist = [100, 50, 65, 82, 23]
thislist.sort(key = myfunc)
print(thislist)

[50, 65, 23, 82, 100]


## Case Insensitive Sort
So if you want a case-insensitive sort function, use str.lower as a key function:



In [95]:
thislist = ["banana", "Orange", "Kiwi", "cherry"]
thislist.sort(key = str.lower)
print(thislist)

['banana', 'cherry', 'Kiwi', 'Orange']


## Reverse Order
What if you want to reverse the order of a list, regardless of the alphabet?

The reverse() method reverses the current sorting order of the elements.

In [96]:
thislist = ["banana", "Orange", "Kiwi", "cherry"]
thislist.reverse()
print(thislist)

['cherry', 'Kiwi', 'Orange', 'banana']


## Copy a List
You cannot copy a list simply by typing list2 = list1, because: list2 will only be a reference to list1, and changes made in list1 will automatically also be made in list2.

There are ways to make a copy, one way is to use the built-in List method copy().

In [97]:
thislist = ["apple", "banana", "cherry"]
mylist = thislist.copy()
print(mylist)

['apple', 'banana', 'cherry']


Another way to make a copy is to use the built-in method list().

In [98]:
mylist = list(thislist)
print(mylist)

['apple', 'banana', 'cherry']


## Join Two Lists
There are several ways to join, or concatenate, two or more lists in Python.

One of the easiest ways are by using the `+` operator.

In [99]:
list1 = ["a", "b", "c"]
list2 = [1, 2, 3]

list3 = list1 + list2
print(list3)

['a', 'b', 'c', 1, 2, 3]


Another way to join two lists is by appending all the items from list2 into list1, one by one:



In [100]:
list1 = ["a", "b" , "c"]
list2 = [1, 2, 3]

for x in list2:
  list1.append(x)

print(list1)

['a', 'b', 'c', 1, 2, 3]


Or you can use the extend() method, which purpose is to add elements from one list to another list:



In [101]:
list1 = ["a", "b" , "c"]
list2 = [1, 2, 3]

list1.extend(list2)
print(list1)

['a', 'b', 'c', 1, 2, 3]


# Tuples

Tuple Methods
Python has two built-in methods that you can use on tuples.

 | Method | Description |
 |--------|-------------|  
 | count() | Returns the number of times a specified value occurs in a tuple | 
 | index() | Searches the tuple for a specified value and returns the position of where it was found | 
 
 
 Tuples are used to store multiple items in a single variable.

Tuple is one of 4 built-in data types in Python used to store collections of data, the other 3 are List, Set, and Dictionary, all with different qualities and usage.

A tuple is a collection which is **ordered** and **unchangeable** and **Allow Duplicates**.

Tuples are written with round brackets.

In [102]:
thistuple = ("apple", "banana", "cherry")
print(thistuple , type(thistuple))

('apple', 'banana', 'cherry') <class 'tuple'>


## Tuple Length

To determine how many items a tuple has, use the len() function:



In [103]:
thistuple = ("apple", "banana", "cherry")
print(len(thistuple))

3


## Create Tuple With One Item
To create a tuple with only one item, you have to add a comma after the item, otherwise Python will not recognize it as a tuple.



In [104]:
thistuple = ("apple",)
print(type(thistuple))

#NOT a tuple
thistuple = ("apple")
print(type(thistuple))

<class 'tuple'>
<class 'str'>


## Tuple Items - Data Types
Tuple items can be of any data type:

In [105]:
tuple1 = ("apple", "banana", "cherry")
tuple2 = (1, 5, 7, 9, 3)
tuple3 = (True, False, False)
tuple4 = ("abc", 34, True, 40, "male")
print(tuple1, tuple2 , tuple3 , tuple4)

('apple', 'banana', 'cherry') (1, 5, 7, 9, 3) (True, False, False) ('abc', 34, True, 40, 'male')


## The tuple() Constructor
It is also possible to use the tuple() constructor to make a tuple.



In [106]:
thistuple = tuple(("apple", "banana", "cherry")) # note the double round-brackets
print(thistuple)

('apple', 'banana', 'cherry')


## Access Tuple Items
You can access tuple items by referring to the index number, inside square brackets:

### Note: The first item has index 0.



In [107]:
thistuple = ("apple", "banana", "cherry")
print(thistuple[1])

banana


## Negative Indexing
Negative indexing means start from the end.

-1 refers to the last item, -2 refers to the second last item etc.

In [108]:
thistuple = ("apple", "banana", "cherry")
print(thistuple[-1])
print(thistuple[-3])

cherry
apple


## Range of Indexes
You can specify a range of indexes by specifying where to start and where to end the range.

When specifying a range, the return value will be a new tuple with the specified items.

Note: The search will start at index 2 (included) and end at index 5 (not included).

Remember that the first item has index 0.



In [109]:
thistuple = ("apple", "banana", "cherry", "orange", "kiwi", "melon", "mango")
print(thistuple[2:5])
print(thistuple[:4])
print(thistuple[2:])
print(thistuple[-4:-1]) # This example returns the items from index -4 (included) to index -1 (excluded)

('cherry', 'orange', 'kiwi')
('apple', 'banana', 'cherry', 'orange')
('cherry', 'orange', 'kiwi', 'melon', 'mango')
('orange', 'kiwi', 'melon')


## Check if Item Exists
To determine if a specified item is present in a tuple use the in keyword:



In [110]:
thistuple = ("apple", "banana", "cherry")
if "apple" in thistuple:
    print("Yes, 'apple' is in the fruits tuple")

Yes, 'apple' is in the fruits tuple


## Change Tuple Values
Once a tuple is created, you cannot change its values. Tuples are unchangeable, or immutable as it also is called.

But **there is a workaround. You can convert the tuple into a list, change the list, and convert the list back into a tuple.**

In [111]:
x = ("apple", "banana", "cherry")
y = list(x)
y[1] = "kiwi"
x = tuple(y)

print(x)

('apple', 'kiwi', 'cherry')


## Add Items
Since tuples are immutable, they do not have a build-in append() method, but there are other ways to add items to a tuple.

1. Convert into a list: Just like the workaround for changing a tuple, you can convert it into a list, add your item(s), and convert it back into a tuple.

In [112]:
thistuple = ("apple", "banana", "cherry")
y = list(thistuple)
y.append("orange")
thistuple = tuple(y)
thistuple

('apple', 'banana', 'cherry', 'orange')

2. Add tuple to a tuple. You are allowed to add tuples to tuples, so if you want to add one item, (or many), create a new tuple with the item(s), and add it to the existing tuple:

Note: When creating a tuple with only one item, remember to include a comma after the item, otherwise it will not be identified as a tuple.

Note: You cannot remove items in a tuple.



In [113]:
thistuple = ("apple", "banana", "cherry")
y = ("orange",)
thistuple += y

print(thistuple)

('apple', 'banana', 'cherry', 'orange')


Tuples are **unchangeable**, so you cannot remove items from it, but you can use the same workaround as we used for changing and adding tuple items:

Or you can delete the tuple completely:



In [114]:
thistuple = ("apple", "banana", "cherry")
y = list(thistuple)
y.remove("apple")
thistuple = tuple(y)
print(thistuple)

del thistuple  # The del keyword can delete the tuple completely
# print(thistuple) this will raise an error because the tuple no longer exists

('banana', 'cherry')


## Unpacking a Tuple
When we create a tuple, we normally assign values to it. This is called "packing" a tuple:

**Note: The number of variables must match the number of values in the tuple, if not, you must use an asterisk to collect the remaining values as a list.**


In [115]:
fruits = ("apple", "banana", "cherry")

(green, yellow, red) = fruits

print(green)
print(yellow)
print(red)

apple
banana
cherry


### Using Asterisk*
If the number of variables is less than the number of values, you can add an * to the variable name and the values will be assigned to the variable as a list:



In [116]:
fruits = ("apple", "banana", "cherry", "strawberry", "raspberry")

(green, yellow, *red) = fruits

print(green)
print(yellow)
print(red)

apple
banana
['cherry', 'strawberry', 'raspberry']


If the asterisk is added to another variable name than the last, Python will assign values to the variable until the number of values left matches the number of variables left.



In [117]:
fruits = ("apple", "mango", "papaya", "pineapple", "cherry")

(green, *tropic, red) = fruits

print(green)
print(red)
print(tropic)


apple
cherry
['mango', 'papaya', 'pineapple']


## Loop Through a Tuple
You can loop through the tuple items by using a for loop.



In [118]:
thistuple = ("apple", "banana", "cherry")
for x in thistuple:
    print(x)

apple
banana
cherry


## Loop Through the Index Numbers
You can also loop through the tuple items by referring to their index number.

Use the range() and len() functions to create a suitable iterable.

In [119]:
thistuple = ("apple", "banana", "cherry")
for i in range(len(thistuple)):
    print(thistuple[i])

apple
banana
cherry


## Using a While Loop
You can loop through the list items by using a while loop.

Use the len() function to determine the length of the tuple, then start at 0 and loop your way through the tuple items by refering to their indexes.

Remember to increase the index by 1 after each iteration.

In [120]:
i = 0
while i < len(thistuple):
    print(thistuple[i])
    i += 1
    

apple
banana
cherry


## Join Two Tuples
To join two or more tuples you can use the + operator:

In [121]:
tuple1 = ("a", "b" , "c")
tuple2 = (1, 2, 3)

tuple3 = tuple1 + tuple2
print(tuple3)

('a', 'b', 'c', 1, 2, 3)


## Multiply Tuples
If you want to multiply the content of a tuple a given number of times, you can use the * operator:

In [122]:
fruits = ("apple", "banana", "cherry")
mytuple = fruits * 3

print(mytuple)

('apple', 'banana', 'cherry', 'apple', 'banana', 'cherry', 'apple', 'banana', 'cherry')


#  Sets
Sets are used to store multiple items in a single variable.

Set is one of 4 built-in data types in Python used to store collections of data, the other 3 are List, Tuple, and Dictionary, all with different qualities and usage.

A set is a collection which is both unordered and unindexed.

Sets are written with curly brackets.


## Set Items
Set items are unordered, unchangeable, and do not allow duplicate values.

## Unordered
Unordered means that the items in a set do not have a defined order.

Set items can appear in a different order every time you use them, and cannot be referred to by index or key.

## Unchangeable
Sets are unchangeable, meaning that we cannot change the items after the set has been created.
## Duplicates Not Allowed


## Set Methods
Python has a set of built-in methods that you can use on sets.

| Method | Description  |
|--------|--------------|
| add() | Adds an element to the set | 
| clear() | Removes all the elements from the set | 
| copy() | Returns a copy of the set | 
| difference() | Returns a set containing the difference between two or more sets | 
| difference_update() | Removes the items in this set that are also included in another, specified set | 
| discard() | Remove the specified item | 
| intersection() | Returns a set, that is the intersection of two other sets | 
| intersection_update() | Removes the items in this set that are not present in other, specified set(s) | 
| isdisjoint() | Returns whether two sets have a intersection or not | 
| issubset() | Returns whether another set contains this set or not | 
| issuperset() | Returns whether this set contains another set or not | 
| pop() | Removes an element from the set | 
| remove() | Removes the specified element | 
| symmetric_difference() | Returns a set with the symmetric differences of two sets | 
| symmetric_difference_update() | inserts the symmetric differences from this set and another | 
| union() | Return a set containing the union of sets | 
| update() | Update the set with the union of this set and others | 

In [123]:
thisset = {"apple", "banana", "cherry"}
print(thisset)

{'cherry', 'apple', 'banana'}


## Get the Length of a Set
To determine how many items a set has, use the len() method.

In [124]:
len(thisset)

3

## Data Types
Set items can be of any data type:

In [125]:
set1 = {"apple", "banana", "cherry"}
set2 = {1, 5, 7, 9, 3}
set3 = {True, False, False}
set4 = {"abc", 34, True, 40, "male"}
type(set4)

set

## The set() Constructor
It is also possible to use the set() constructor to make a set.

In [126]:
thisset = set(("apple", "banana", "cherry")) # note the double round-brackets
print(thisset)

{'cherry', 'apple', 'banana'}


## Access Items
You cannot access items in a set by referring to an index or a key.

But you can loop through the set items using a for loop, or ask if a specified value is present in a set, by using the in keyword.

### Change Items

Once a set is created, you cannot change its items, but you can add new items.



In [127]:
thisset = {"apple", "banana", "cherry"}

for x in thisset:
    print(x)
    
    
print("banana" in thisset)

cherry
apple
banana
True


## Add Items
Once a set is created, you cannot change its items, but you can add new items.

To add one item to a set use the add() method.

In [128]:
thisset = {"apple", "banana", "cherry"}

thisset.add("orange")

print(thisset)

{'cherry', 'apple', 'banana', 'orange'}


## Add Sets
To add items from another set into the current set, use the update() method.

In [129]:
thisset = {"apple", "banana", "cherry"}
tropical = {"pineapple", "mango", "papaya"}

thisset.update(tropical)

print(thisset)

{'mango', 'banana', 'cherry', 'apple', 'pineapple', 'papaya'}


## Add Any Iterable
The object in the update() method does not have to be a set, it can be any iterable object (tuples, lists, dictionaries etc.).

In [130]:
thisset = {"apple", "banana", "cherry"}
mylist = ["kiwi", True]

thisset.update(mylist)

print(thisset)

{True, 'kiwi', 'banana', 'apple', 'cherry'}


## Remove Item
To remove an item in a set, use the remove(), or the discard() method.

In [131]:
thisset = {"apple", "banana", "cherry"}

thisset.remove("banana")

print(thisset)

# Note: If the item to remove does not exist, remove() will raise an error.


# thisset.remove("Ali") 



{'cherry', 'apple'}


In [132]:
thisset = {"apple", "banana", "cherry"}

thisset.discard("banana")

print(thisset)

{'cherry', 'apple'}


You can also use the pop() method to remove an item, but this method will remove the last item. Remember that sets are unordered, so you will not know what item that gets removed.

The return value of the pop() method is the removed item.

In [133]:
thisset = {"apple", "banana", "cherry"}


# Note: Sets are unordered, so when using the pop() method, you do not know which item that gets removed.

x = thisset.pop()

print(x)

print(thisset)

cherry
{'apple', 'banana'}


In [134]:
thisset = {"apple", "banana", "cherry"}

thisset.clear()

print(thisset)

set()


In [135]:
# The del keyword will delete the set completely:

thisset = {"apple", "banana", "cherry"}

del thisset

# print(thisset)  NameError: name 'thisset' is not defined

## Loop Items
You can loop through the set items by using a for loop:

In [136]:
thisset = {"apple", "banana", "cherry"}

for x in thisset:
  print(x)

cherry
apple
banana


## Join Two Sets
There are several ways to join two or more sets in Python.

You can use the union() method that returns a new set containing all items from both sets, or the update() method that inserts all the items from one set into another:

In [137]:
set1 = {"a", "b" , "c"}
set2 = {1, 2, 3}

set3 = set1.union(set2)
print(set3)

set1.update(set2)
print(set1)

{1, 2, 3, 'c', 'b', 'a'}
{1, 2, 3, 'c', 'b', 'a'}


## Keep ONLY the Duplicates
The intersection_update() method will keep only the items that are present in both sets.

In [138]:
x = {"apple", "banana", "cherry"}
y = {"google", "microsoft", "apple"}

x.intersection_update(y)

print(x)

{'apple'}


In [139]:
# The intersection() method will return a new set, that only contains the items that are present in both sets.
x = {"apple", "banana", "cherry"}
y = {"google", "microsoft", "apple"}

z = x.intersection(y)

print(z)


{'apple'}


## Keep All, But NOT the Duplicates
The symmetric_difference_update() method will keep only the elements that are NOT present in both sets.

In [140]:
x = {"apple", "banana", "cherry"}
y = {"google", "microsoft", "apple"}

x.symmetric_difference_update(y)

print(x)

{'banana', 'microsoft', 'google', 'cherry'}


In [141]:
 #The symmetric_difference() method will return a new set, that contains only the elements that are NOT present in both sets.

x = {"apple", "banana", "cherry"}
y = {"google", "microsoft", "apple"}

z = x.symmetric_difference(y)

print(z)

{'banana', 'microsoft', 'google', 'cherry'}


# Dictionaries

Dictionaries are used to store data values in key:value pairs.

A dictionary is a collection which is ordered*, changeable and does not allow duplicates.

As of Python version 3.7, dictionaries are ordered. In Python 3.6 and earlier, dictionaries are unordered.

## Dictionary Items
Dictionary items are ordered, changeable, and does not allow duplicates.

Dictionary items are presented in key:value pairs, and can be referred to by using the key name.

## Ordered or Unordered?
As of Python version 3.7, dictionaries are ordered. In Python 3.6 and earlier, dictionaries are unordered.

When we say that dictionaries are ordered, it means that the items have a defined order, and that order will not change.

Unordered means that the items does not have a defined order, you cannot refer to an item by using an index.

## Changeable
Dictionaries are changeable, meaning that we can change, add or remove items after the dictionary has been created.

## Duplicates Not Allowed

## Dictionary Methods
Python has a set of built-in methods that you can use on dictionaries.

| Method	| Description |
|-----------|-------------|
| clear()	| Removes all the elements from the dictionary |
| copy()	| Returns a copy of the dictionary |
| fromkeys()	| Returns a dictionary with the specified keys and value |
| get()	| Returns the value of the specified key |
| items()	| Returns a list containing a tuple for each key value pair |
| keys()	| Returns a list containing the dictionary's keys |
| pop()	| Removes the element with the specified key |
| popitem()	| Removes the last inserted key-value pair |
| setdefault()	| Returns the value of the specified key. If the key does not exist: insert the key, with the specified value |
| update()	| Updates the dictionary with the specified key-value pairs |
| values()	| Returns a list of all the values in the dictionary |

In [142]:
thisdict = {
  "brand": "Ford",
  "electric": False,
  "year": 1964,
  "colors": ["red", "white", "blue"]
}
print(thisdict)
print(thisdict["brand"])


{'brand': 'Ford', 'electric': False, 'year': 1964, 'colors': ['red', 'white', 'blue']}
Ford


## Dictionary Length
To determine how many items a dictionary has, use the len() function:

## Dictionary Items - Data Types
The values in dictionary items can be of any data type:

In [143]:
print(len(thisdict))
print(type(thisdict))


4
<class 'dict'>


## Accessing Items
You can access the items of a dictionary by referring to its key name, inside square brackets:

There is also a method called get() that will give you the same result:



In [144]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
x = thisdict["model"]
print(x)
x = thisdict.get("year")
print(x)

Mustang
1964


## Get Keys
The keys() method will return a list of all the keys in the dictionary.
The list of the keys is a view of the dictionary, meaning that any changes done to the dictionary will be reflected in the keys list.


In [145]:
x = thisdict.keys()
x 

dict_keys(['brand', 'model', 'year'])

In [146]:
car = {
"brand": "Ford",
"model": "Mustang",
"year": 1964
}

x = car.keys()

print(x) #before the change

car["color"] = "white"

print(x) #after the change

dict_keys(['brand', 'model', 'year'])
dict_keys(['brand', 'model', 'year', 'color'])


## Change Values
You can change the value of a specific item by referring to its key name:

In [147]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict["year"] = 2018
thisdict

{'brand': 'Ford', 'model': 'Mustang', 'year': 2018}

## Update Dictionary
The update() method will update the dictionary with the items from the given argument.

The argument must be a dictionary, or an iterable object with key:value pairs.

In [148]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict.update({"year": 2020 , "brand": "Toyota"})

thisdict

{'brand': 'Toyota', 'model': 'Mustang', 'year': 2020}

## Adding Items
Adding an item to the dictionary is done by using a new index key and assigning a value to it:

In [149]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict["color"] = "red"

print(thisdict)

{'brand': 'Ford', 'model': 'Mustang', 'year': 1964, 'color': 'red'}


In [150]:
thisdict.update({"Owner": "Ali"})
thisdict

{'brand': 'Ford',
 'model': 'Mustang',
 'year': 1964,
 'color': 'red',
 'Owner': 'Ali'}

## Removing Items
There are several methods to remove items from a dictionary:

In [151]:
# The pop() method removes the item with the specified key name:

thisdict = {
    "Owner" : "Ali",
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964 ,
    
}
thisdict.pop("model")
print(thisdict)

thisdict.popitem() # The popitem() method removes the last inserted item
print(thisdict)

# The del keyword removes the item with the specified key name:

del thisdict["brand"]
print(thisdict)

del thisdict
# print(thisdict) this will cause an error because "thisdict" no longer exists.

{'Owner': 'Ali', 'brand': 'Ford', 'year': 1964}
{'Owner': 'Ali', 'brand': 'Ford'}
{'Owner': 'Ali'}


In [152]:
## The clear() method empties the dictionary:

thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict.clear()
print(thisdict)

{}


## Loop Through a Dictionary
You can loop through a dictionary by using a for loop.

When looping through a dictionary, the return value are the keys of the dictionary, but there are methods to return the values as well.

In [153]:
thisdict = {
    "Owner" : "Ali",
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964 ,
    
}

# Print all key names in the dictionary, one by one:


for x in thisdict:
    print(x)
    

Owner
brand
model
year


In [154]:
# Print all values in the dictionary, one by one:


for x in thisdict:
    print(thisdict[x])

Ali
Ford
Mustang
1964


In [155]:
# You can also use the values() method to return values of a dictionary:


for x in thisdict.values():
    print(x)

Ali
Ford
Mustang
1964


In [156]:
# You can use the keys() method to return the keys of a dictionary:


for x in thisdict.keys():
  print(x)

Owner
brand
model
year


In [157]:
# Loop through both keys and values, by using the items() method:

for x, y in thisdict.items():
  print(x, y)

Owner Ali
brand Ford
model Mustang
year 1964


## Copy a Dictionary
You cannot copy a dictionary simply by typing dict2 = dict1, because: dict2 will only be a reference to dict1, and changes made in dict1 will automatically also be made in dict2.

There are ways to make a copy, one way is to use the built-in Dictionary method copy().



In [158]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
mydict = thisdict.copy()
print(mydict)

{'brand': 'Ford', 'model': 'Mustang', 'year': 1964}


In [159]:
# Another way to make a copy is to use the built-in function dict().

thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
mydict = dict(thisdict)
print(mydict)

{'brand': 'Ford', 'model': 'Mustang', 'year': 1964}


## Nested Dictionaries
A dictionary can contain dictionaries, this is called nested dictionaries.

In [160]:
myfamily = {
  "child1" : {
    "name" : "Emil",
    "year" : 2004
  },
  "child2" : {
    "name" : "Tobias",
    "year" : 2007
  },
  "child3" : {
    "name" : "Linus",
    "year" : 2011
  }
}

myfamily , myfamily['child2']["year"]

({'child1': {'name': 'Emil', 'year': 2004},
  'child2': {'name': 'Tobias', 'year': 2007},
  'child3': {'name': 'Linus', 'year': 2011}},
 2007)

In [161]:
# Or, if you want to add three dictionaries into a new dictionary:

child1 = {
  "name" : "Emil",
  "year" : 2004
}
child2 = {
  "name" : "Tobias",
  "year" : 2007
}
child3 = {
  "name" : "Linus",
  "year" : 2011
}

myfamily = {
  "child1" : child1,
  "child2" : child2,
  "child3" : child3
}

myfamily , myfamily['child2']["name"]

({'child1': {'name': 'Emil', 'year': 2004},
  'child2': {'name': 'Tobias', 'year': 2007},
  'child3': {'name': 'Linus', 'year': 2011}},
 'Tobias')

# Python If ... Else
Python supports the usual logical conditions from mathematics:

- Equals: a == b
- Not Equals: a != b
- Less than: a < b
- Less than or equal to: a <= b
- Greater than: a > b
- Greater than or equal to: a >= b

These conditions can be used in several ways, most commonly in "if statements" and loops.

An "if statement" is written by using the if keyword.

In [162]:
a = 33
b = 200
if b > a:
  print("b is greater than a")

# In this example we use two variables, a and b,
# which are used as part of the if statement to test whether b is greater than a. 
# As a is 33, and b is 200, we know that 200 is greater than 33, and so we print to screen that "b is greater than a".

b is greater than a


## Elif
The elif keyword is pythons way of saying "if the previous conditions were not true, then try this condition".

In [163]:
a = 33
b = 33
if b > a :
    print("b is greater than a")
elif a == b:
    print("a and b are equal")
    
# In this example a is equal to b, so the first condition is not true, but the elif condition is true,
# so we print to screen that "a and b are equal".


a and b are equal


## Else
The else keyword catches anything which isn't caught by the preceding conditions.

In [164]:
a = 200
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")
else:
  print("a is greater than b")

a is greater than b


In [165]:
a = 200
b = 33
if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

b is not greater than a


In [166]:
a = 2
b = 330
print("A") if a > b else print("B")

# This technique is known as Ternary Operators, or Conditional Expressions.

B


In [167]:
print("A") if a > b else print("=") if a == b else print("B")

B


In [168]:
# Test if a is greater than b, AND if c is greater than a:

a = 200
b = 33
c = 500
if a > b and c > a:
  print("Both conditions are True")

Both conditions are True


In [169]:
# Test if a is greater than b, OR if a is greater than c:
a = 200
b = 33
c = 500
if a > b or a > c:
  print("At least one of the conditions is True")

At least one of the conditions is True


## Nested If
You can have if statements inside if statements, this is called nested if statements.

In [170]:
x = 41

if x > 10:
  print("Above ten,")
  if x > 20:
    print("and also above 20!")
  else:
    print("but not above 20.")

Above ten,
and also above 20!


## The pass Statement
if statements cannot be empty, but if you for some reason have an if statement with no content, put in the pass statement to avoid getting an error.

In [171]:
a = 33
b = 200

if b > a:
    pass

# While Loops

Python has two primitive loop commands:

- while loops

- for loops

The while Loop

With the `while` loop we can execute a set of statements as long as a condition is true.

Note: remember to increment i, or else the loop will continue forever.

The while loop requires relevant variables to be ready, in this example we need to define an indexing variable, i, which we set to 1.



In [172]:
i = 1
while i < 6:
    print(i)
    i +=1

1
2
3
4
5


## The break Statement
With the break statement we can stop the loop even if the while condition is true:

In [173]:
i = 50
while i < 100:
    print(i)
    if  i == 53:
        break
    i +=1

50
51
52
53


## The continue Statement
With the continue statement we can stop the current iteration, and continue with the next:

In [174]:
i = 0
while i < 10:
    i +=1
    if i % 2:
        continue
    print(i)

2
4
6
8
10


In [175]:
i = 1
while i < 6:
    print(i)
    i += 1
else:
    print("i is no longer less than 6")

1
2
3
4
5
i is no longer less than 6


# Python For Loops
A for loop is used for iterating over a sequence (that is either a list, a tuple, a dictionary, a set, or a string).

This is less like the for keyword in other programming languages, and works more like an iterator method as found in other object-orientated programming languages.

With the for loop we can execute a set of statements, once for each item in a list, tuple, set etc.

In [176]:
fruits =["apple" , "cherry"  , "HOLO"]

for i in fruits:
    print(i)

apple
cherry
HOLO


## Looping Through a String
Even strings are iterable objects, they contain a sequence of characters:

In [177]:
for i in "Soodabeh":
    print(i)

S
o
o
d
a
b
e
h


## The break Statement
With the break statement we can stop the loop before it has looped through all the items:

In [178]:
fruits =["apple" , "cherry"  , "HOLO" , "Peach"]

for i in fruits:
    print(i)
    if i == "HOLO":
        break

apple
cherry
HOLO


In [179]:
fruits =["apple" , "cherry"  , "HOLO" , "Peach"]

for i in fruits:
#     print(i)
    if i == "HOLO":
        break
    print(i)

apple
cherry


## The continue Statement
With the continue statement we can stop the current iteration of the loop, and continue with the next:

In [180]:
fruits = ["apple", "banana", "cherry"]
for x in fruits:
  if x == "banana":
    continue
  print(x)

apple
cherry


## The range() Function
To loop through a set of code a specified number of times, we can use the range() function,

The range() function returns a sequence of numbers, starting from 0 by default, and increments by 1 (by default), and ends at a specified number.

**Note that range(6) is not the values of 0 to 6, but the values 0 to 5.**

The range() function defaults to 0 as a starting value, however it is possible to specify the starting value by adding a parameter: range(2, 6), which means values from 2 to 6 (but not including 6):

In [181]:
for x in range(6):
    print(x)

0
1
2
3
4
5


In [182]:
for x in range(2, 6):
  print(x)

2
3
4
5


The range() function defaults to increment the sequence by 1, however it is possible to specify the increment value by adding a third parameter: range(2, 30, 3):



In [183]:
for x in range(2, 30, 6):
  print(x)

2
8
14
20
26


## Else in For Loop
The else keyword in a for loop specifies a block of code to be executed when the loop is finished:

**Note: The else block will NOT be executed if the loop is stopped by a break statement.**

In [184]:
for x in range(6):
    print(x)
else:
    print("Finally finished!")

0
1
2
3
4
5
Finally finished!


## Nested Loops
A nested loop is a loop inside a loop.

The "inner loop" will be executed one time for each iteration of the "outer loop":

In [185]:
adj = ["red", "big", "tasty"]
fruits = ["apple", "banana"]

for x in adj:
  for y in fruits:
    print(x, y)

red apple
red banana
big apple
big banana
tasty apple
tasty banana


## The pass Statement
for loops cannot be empty, but if you for some reason have a for loop with no content, put in the pass statement to avoid getting an error.

In [186]:
for x in [0, 1, 2]:
      pass

# Python Functions
A function is a block of code which only runs when it is called.

You can pass data, known as parameters, into a function.

A function can return data as a result.

## Creating a Function
In Python a function is defined using the def keyword:

In [187]:
def my_function():
  print("Hello from a function")

my_function()


Hello from a function


## Arguments
Information can be passed into functions as arguments.

Arguments are specified after the function name, inside the parentheses. You can add as many arguments as you want, just separate them with a comma.

The following example has a function with one argument (fname). When the function is called, we pass along a first name, which is used inside the function to print the full name:

In [188]:
def my_function(fname):
  print(fname + " Refsnes")

my_function("Emil")
my_function("Tobias")
my_function("Linus")

Emil Refsnes
Tobias Refsnes
Linus Refsnes


Arguments are often shortened to args in Python documentations.

## Parameters or Arguments?
The terms parameter and argument can be used for the same thing: information that are passed into a function.

From a function's perspective:

A parameter is the variable listed inside the parentheses in the function definition.

An argument is the value that is sent to the function when it is called.

## Number of Arguments
By default, a function must be called with the correct number of arguments. Meaning that if your function expects 2 arguments, you have to call the function with 2 arguments, not more, and not less.

In [189]:
def my_function(fname, lname):
  print(fname + " " + lname)

my_function("Emil", "Refsnes")
# my_function("Emil") TypeError: my_function() missing 1 required positional argument: 'lname'

Emil Refsnes


## Arbitrary Arguments, *args
If you do not know how many arguments that will be passed into your function, add a * before the parameter name in the function definition.

This way the function will receive a tuple of arguments, and can access the items accordingly:

**Arbitrary Arguments are often shortened to *args in Python documentations.**

In [190]:
def my_function(*kids):
  print("The youngest child is " + kids[2])

my_function("Emil", "Tobias", "Linus")


The youngest child is Linus


## Keyword Arguments
You can also send arguments with the key = value syntax.

This way the order of the arguments does not matter.

In [191]:
def my_function(child3, child2, child1):
  print("The youngest child is " + child3)

my_function(child1 = "Emil", child2 = "Tobias", child3 = "Linus")


The youngest child is Linus


The phrase Keyword Arguments are often shortened to kwargs in Python documentations.

## Arbitrary Keyword Arguments, **kwargs
If you do not know how many keyword arguments that will be passed into your function, add two asterisk: ** before the parameter name in the function definition.

This way the function will receive a dictionary of arguments, and can access the items accordingly:

In [192]:
def my_function(**kid):
    print("His last name is " + kid["lname"])

my_function(fname = "Tobias", lname = "Refsnes")

His last name is Refsnes


Arbitrary Kword Arguments are often shortened to **kwargs in Python documentations.

## Default Parameter Value
The following example shows how to use a default parameter value.

If we call the function without argument, it uses the default value:

In [193]:
def my_function(country = "Norway"):
  print("I am from " + country)

my_function("Sweden")
my_function("India")
my_function()
my_function("Brazil")

I am from Sweden
I am from India
I am from Norway
I am from Brazil


## Passing a List as an Argument
You can send any data types of argument to a function (string, number, list, dictionary etc.), and it will be treated as the same data type inside the function.

E.g. if you send a List as an argument, it will still be a List when it reaches the function:

In [194]:
def my_function(food):
    for x in food:
        print(x)

fruits = ["apple", "banana", "cherry"]

my_function(fruits)

apple
banana
cherry


## Return Values
To let a function return a value, use the return statement:

In [195]:
def my_function(x):
  return 5 * x

print(my_function(3))
print(my_function(5))
print(my_function(9))

15
25
45


## The pass Statement
function definitions cannot be empty, but if you for some reason have a function definition with no content, put in the pass statement to avoid getting an error.

In [196]:
def myfunction():
      pass

## Recursion
Python also accepts function recursion, which means a defined function can call itself.

Recursion is a common mathematical and programming concept. It means that a function calls itself. This has the benefit of meaning that you can loop through data to reach a result.

The developer should be very careful with recursion as it can be quite easy to slip into writing a function which never terminates, or one that uses excess amounts of memory or processor power. However, when written correctly recursion can be a very efficient and mathematically-elegant approach to programming.

In this example, tri_recursion() is a function that we have defined to call itself ("recurse"). We use the k variable as the data, which decrements (-1) every time we recurse. The recursion ends when the condition is not greater than 0 (i.e. when it is 0).

To a new developer it can take some time to work out how exactly this works, best way to find out is by testing and modifying it.

In [197]:
def tri_recursion(k):
  if(k > 0):
    result = k + tri_recursion(k - 1)
    print(result)
  else:
    result = 0
  return result

print("\n\nRecursion Example Results")
tri_recursion(6)



Recursion Example Results
1
3
6
10
15
21


21

# Lambda
A lambda function is a small anonymous function.

A lambda function can take any number of arguments, but can only have one expression.

Syntax

```sql
lambda arguments : expression
```

In [198]:
x = lambda a : a + 10
print(x(5))

x = lambda a, b : a * b
print(x(5, 6))

x = lambda a, b, c : a + b + c
print(x(5, 6, 2))

15
30
13


## Why Use Lambda Functions?
The power of lambda is better shown when you use them as an anonymous function inside another function.

Say you have a function definition that takes one argument, and that argument will be multiplied with an unknown number:

In [199]:
def myfunc(n):
    return lambda a : a * n

mydoubler = myfunc(2)

print(mydoubler(11))

22


In [200]:
def myfunc(n):
  return lambda a : a * n

mytripler = myfunc(3)

print(mytripler(11))

33


In [201]:
def myfunc(n):
  return lambda a : a * n

mydoubler = myfunc(2)
mytripler = myfunc(3)

print(mydoubler(11))
print(mytripler(11))

22
33


# Classes/Objects
Python is an object oriented programming language.

Almost everything in Python is an object, with its properties and methods.

A Class is like an object constructor, or a "blueprint" for creating objects.

## Create a Class
To create a class, use the keyword class:

In [202]:
class MyClass:
  x = 5


## Create Object
Now we can use the class named MyClass to create objects:

In [203]:
p1 = MyClass()
print(p1.x)

5


## The __init__() Function
The examples above are classes and objects in their simplest form, and are not really useful in real life applications.

To understand the meaning of classes we have to understand the built-in __init__() function.

All classes have a function called __init__(), which is always executed when the class is being initiated.

Use the __init__() function to assign values to object properties, or other operations that are necessary to do when the object is being created:


**Note: The __init__() function is called automatically every time the class is being used to create a new object.**

In [204]:
class Person:
  def __init__(self, name, age):
    self.name = name
    self.age = age

p1 = Person("John", 36)

print(p1.name)
print(p1.age)

John
36


## Object Methods
Objects can also contain methods. Methods in objects are functions that belong to the object.

Let us create a method in the Person class:

**Note: The self parameter is a reference to the current instance of the class, and is used to access variables that belong to the class.**

In [205]:
class Person:
  def __init__(self, name, age):
    self.name = name
    self.age = age

  def myfunc(self):
    print("Hello my name is " + self.name)

p1 = Person("John", 36)
p1.myfunc()

Hello my name is John


## The self Parameter
The self parameter is a reference to the current instance of the class, and is used to access variables that belongs to the class.

It does not have to be named self , you can call it whatever you like, but it has to be the first parameter of any function in the class:

In [206]:
class Person:
  def __init__(mysillyobject, name, age):
    mysillyobject.name = name
    mysillyobject.age = age

  def myfunc(abc):
    print("Hello my name is " + abc.name)

p1 = Person("John", 36)
p1.myfunc()

Hello my name is John


## Modify Object Properties
You can modify properties on objects like this:

In [207]:
p1.age = 40

## Delete Object Properties
You can delete properties on objects by using the del keyword:

In [208]:
del p1.age


## Delete Objects
You can delete objects by using the del keyword:

In [209]:
del p1


## The pass Statement
class definitions cannot be empty, but if you for some reason have a class definition with no content, put in the pass statement to avoid getting an error.

In [210]:
class Person:
  pass

## Inheritance
Inheritance allows us to define a class that inherits all the methods and properties from another class.

Parent class is the class being inherited from, also called base class.

Child class is the class that inherits from another class, also called derived class.

## Create a Parent Class
Any class can be a parent class, so the syntax is the same as creating any other class:

In [211]:
class Person:
  def __init__(self, fname, lname):
    self.firstname = fname
    self.lastname = lname

  def printname(self):
    print(self.firstname, self.lastname)

#Use the Person class to create an object, and then execute the printname method:

x = Person("John", "Doe")
x.printname()

John Doe


## Create a Child Class
To create a class that inherits the functionality from another class, send the parent class as a parameter when creating the child class:

In [212]:
class Student(Person):
      pass

In [213]:
x = Student("Mike", "Olsen")
x.printname()

Mike Olsen


## Add the __init__() Function
So far we have created a child class that inherits the properties and methods from its parent.

We want to add the __init__() function to the child class (instead of the pass keyword).

Note: The __init__() function is called automatically every time the class is being used to create a new object.

When you add the __init__() function, the child class will no longer inherit the parent's __init__() function.

Note: The child's __init__() function overrides the inheritance of the parent's __init__() function.

To keep the inheritance of the parent's __init__() function, add a call to the parent's __init__() function:

In [214]:
class Student(Person):
  def __init__(self, fname, lname):
    #add properties etc.
        Person.__init__(self, fname, lname)


## Use the super() Function
Python also has a super() function that will make the child class inherit all the methods and properties from its parent:

In [215]:
class Student(Person):
  def __init__(self, fname, lname):
    super().__init__(fname, lname)

## Add Properties
In the example below, the year 2019 should be a variable, and passed into the Student class when creating student objects. To do so, add another parameter in the __init__() function:



In [216]:
class Student(Person):
  def __init__(self, fname, lname):
    super().__init__(fname, lname)
    self.graduationyear = 2019

In [217]:
class Student(Person):
  def __init__(self, fname, lname, year):
    super().__init__(fname, lname)
    self.graduationyear = year

x = Student("Mike", "Olsen", 2019)

## Add Methods
If you add a method in the child class with the same name as a function in the parent class, the inheritance of the parent method will be overridden.



In [218]:
class Student(Person):
  def __init__(self, fname, lname, year):
    super().__init__(fname, lname)
    self.graduationyear = year

  def welcome(self):
    print("Welcome", self.firstname, self.lastname, "to the class of", self.graduationyear)

# Iterators
An iterator is an object that contains a countable number of values.

An iterator is an object that can be iterated upon, meaning that you can traverse through all the values.

Technically, in Python, an iterator is an object which implements the iterator protocol, which consist of the methods __iter__() and __next__().

## Iterator vs Iterable
Lists, tuples, dictionaries, and sets are all iterable objects. They are iterable containers which you can get an iterator from.

All these objects have a iter() method which is used to get an iterator:

In [219]:
mytuple = ("apple", "banana", "cherry")
myit = iter(mytuple)

print(next(myit))
print(next(myit))
print(next(myit))

apple
banana
cherry


In [220]:
mystr = "banana"
myit = iter(mystr)

print(next(myit))
print(next(myit))
print(next(myit))
print(next(myit))
print(next(myit))
print(next(myit))

b
a
n
a
n
a


## Looping Through an Iterator
We can also use a for loop to iterate through an iterable object:

In [221]:
mytuple = ("apple", "banana", "cherry")

for x in mytuple:
  print(x)

for x in mystr:
  print(x)

apple
banana
cherry
b
a
n
a
n
a


he for loop actually creates an iterator object and executes the next() method for each loop.

## Create an Iterator
To create an object/class as an iterator you have to implement the methods __iter__() and __next__() to your object.

As you have learned in the Python Classes/Objects chapter, all classes have a function called __init__(), which allows you to do some initializing when the object is being created.

The __iter__() method acts similar, you can do operations (initializing etc.), but must always return the iterator object itself.

The __next__() method also allows you to do operations, and must return the next item in the sequence.

In [222]:
class MyNumbers:
  def __iter__(self):
    self.a = 1
    return self

  def __next__(self):
    x = self.a
    self.a += 1
    return x

myclass = MyNumbers()
myiter = iter(myclass)

print(next(myiter))
print(next(myiter))
print(next(myiter))
print(next(myiter))
print(next(myiter))

1
2
3
4
5


## StopIteration
The example above would continue forever if you had enough next() statements, or if it was used in a for loop.

To prevent the iteration to go on forever, we can use the StopIteration statement.

In the __next__() method, we can add a terminating condition to raise an error if the iteration is done a specified number of times:

In [223]:
class MyNumbers:
  def __iter__(self):
    self.a = 1
    return self

  def __next__(self):
    if self.a <= 20:
      x = self.a
      self.a += 1
      return x
    else:
      raise StopIteration

myclass = MyNumbers()
myiter = iter(myclass)

for x in myiter:
  print(x)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20


# Scope
A variable is only available from inside the region it is created. This is called scope.

## Local Scope
A variable created inside a function belongs to the local scope of that function, and can only be used inside that function.

In [224]:
def myfunc():
  x = 300
  print(x)

myfunc() 

300


## Function Inside Function
As explained in the example above, the variable x is not available outside the function, but it is available for any function inside the function:

In [225]:
def myfunc():
  x = 300
  def myinnerfunc():
    print(x)
  myinnerfunc()

myfunc()

300


## Global Scope
A variable created in the main body of the Python code is a global variable and belongs to the global scope.

Global variables are available from within any scope, global and local.

In [226]:
x = 2000

def myfunc():
  print(x)

myfunc()

print(x)

2000
2000


## Naming Variables
If you operate with the same variable name inside and outside of a function, Python will treat them as two separate variables, one available in the global scope (outside the function) and one available in the local scope (inside the function):

In [227]:
x = 300

def myfunc():
  x = 200
  print(x)

myfunc()

print(x)

200
300


## Global Keyword
If you need to create a global variable, but are stuck in the local scope, you can use the global keyword.

The global keyword makes the variable global.

**Also, use the global keyword if you want to make a change to a global variable inside a function.**

In [228]:
def myfunc():
  global x
  x = 300

myfunc()

print(x)

300


In [229]:
x = 300

def myfunc():
  global x
  x = 200

myfunc()

print(x)

200


https://www.w3schools.com/python/python_modules.asp

# Modules
Consider a module to be the same as a code library.

A file containing a set of functions you want to include in your application.

## Create a Module
To create a module just save the code you want in a file with the file extension .py:

## Use a Module
Now we can use the module we just created, by using the import statement:

**Note: When using a function from a module, use the syntax: module_name.function_name.**

In [2]:
import mymodule

mymodule.greeting("Jonathan")

Hello, Jonathan


## Variables in Module
The module can contain functions, as already described, but also variables of all types (arrays, dictionaries, objects etc):

## Naming a Module
You can name the module file whatever you like, but it must have the file extension .py

##  Re-naming a Module
You can create an alias when you import a module, by using the as keyword:

In [5]:
import mymodule as mx

a = mx.person1["age"]
print(a)

36


## Built-in Modules
There are several built-in modules in Python, which you can import whenever you like.

In [7]:
import platform

x = platform.system()
print(x)

Windows


## Using the dir() Function
There is a built-in function to list all the function names (or variable names) in a module. The dir() function:

Note: The dir() function can be used on all modules, also the ones you create yourself.



In [8]:
import platform

x = dir(platform)
print(x)

['DEV_NULL', '_UNIXCONFDIR', '_WIN32_CLIENT_RELEASES', '_WIN32_SERVER_RELEASES', '__builtins__', '__cached__', '__copyright__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '__version__', '_comparable_version', '_component_re', '_default_architecture', '_dist_try_harder', '_follow_symlinks', '_ironpython26_sys_version_parser', '_ironpython_sys_version_parser', '_java_getprop', '_libc_search', '_linux_distribution', '_lsb_release_version', '_mac_ver_xml', '_node', '_norm_version', '_parse_release_file', '_platform', '_platform_cache', '_pypy_sys_version_parser', '_release_filename', '_release_version', '_supported_dists', '_sys_version', '_sys_version_cache', '_sys_version_parser', '_syscmd_file', '_syscmd_uname', '_syscmd_ver', '_uname_cache', '_ver_output', '_ver_stages', 'architecture', 'collections', 'dist', 'java_ver', 'libc_ver', 'linux_distribution', 'mac_ver', 'machine', 'node', 'os', 'platform', 'popen', 'processor', 'python_branch', 'python_build

## Import From Module
You can choose to import only parts from a module, by using the from keyword.

In [10]:
from mymodule import person1

print (person1["age"])

36


# Datetime

A date in Python is not a data type of its own, but we can import a module named datetime to work with dates as date objects.



In [20]:
import datetime

x = datetime.datetime.now()
print(x)

2021-08-03 11:53:50.445027


## Date Output
When we execute the code from the example above the result will be:

2021-08-03 11:53:50.445027

The date contains year, month, day, hour, minute, second, and microsecond.

The datetime module has many methods to return information about the date object.

Here are a few examples, you will learn more about them later in this chapter:

In [22]:
import datetime
x =  datetime.datetime.now()
print(x)
x.day , x.year , x.second

2021-08-03 11:54:55.217769


(3, 2021, 55)

## Creating Date Objects
To create a date, we can use the datetime() class (constructor) of the datetime module.

The datetime() class requires three parameters to create a date: year, month, day.

The datetime() class also takes parameters for time and timezone (`hour`, `minute`, `second`, `microsecond`, `tzone`), but they are optional, and has a default value of 0, (None for timezone).



In [28]:
x = datetime.datetime(2020 , 5 , 17 , 5 , 35 , 35)
print(x)


2020-05-17 05:35:35


## The strftime() Method
The datetime object has a method for formatting date objects into readable strings.

The method is called strftime(), and takes one parameter, format, to specify the format of the returned string:

In [30]:
x = datetime.datetime(2018, 6, 1)

print(x.strftime("%B"))

June


# Math

Python has a set of built-in math functions, including an extensive math module, that allows you to perform mathematical tasks on numbers.

Built-in Math Functions
The min() and max() functions can be used to find the lowest or highest value in an iterable:

The math.ceil() method rounds a number upwards to its nearest integer, and the math.floor() method rounds a number downwards to its nearest integer, and returns the result:

The math.pi constant, returns the value of PI (3.14...):



In [32]:
x = min(5, 10, 25)
y = max(5, 10, 25)

print(x)
print(y)

5
25


In [35]:
abs(-7.25) , pow(4, 3)

(7.25, 64)

In [40]:
import math
math.sqrt(64) # The math.sqrt() method for example, returns the square root of a number:

8.0

In [42]:
 math.ceil(1.4) , math.floor(1.4) , math.pi

(2, 1, 3.141592653589793)

# JSON
SON is a syntax for storing and exchanging data.

JSON is text, written with JavaScript object notation.

## JSON in Python
Python has a built-in package called json, which can be used to work with JSON data.

## Parse JSON - Convert from JSON to Python
If you have a JSON string, you can parse it by using the json.loads() method.

The result will be a Python dictionary.

In [45]:
import json

# some JSON:
x =  '{ "name":"John", "age":30, "city":"New York"}'

# parse x:
y = json.loads(x)

# the result is a Python dictionary:
print(y["age"])

30


## Convert from Python to JSON
If you have a Python object, you can convert it into a JSON string by using the json.dumps() method.

You can convert Python objects of the following types, into JSON strings:

- dict
- list
- tuple
- string
- int
- float
- True
- False
- None

In [47]:
import json

# a Python object (dict):
x = {
  "name": "John",
  "age": 30,
  "city": "New York"
}

# convert into JSON:
y = json.dumps(x)

# the result is a JSON string:
print(y)

{"name": "John", "age": 30, "city": "New York"}


In [49]:
import json

print(json.dumps({"name": "John", "age": 30}))
print(json.dumps(["apple", "bananas"]))
print(json.dumps(("apple", "bananas")))
print(json.dumps("hello"))
print(json.dumps(42))
print(json.dumps(31.76))
print(json.dumps(True))
print(json.dumps(False))
print(json.dumps(None))

{"name": "John", "age": 30}
["apple", "bananas"]
["apple", "bananas"]
"hello"
42
31.76
true
false
null


In [50]:
import json

x = {
  "name": "John",
  "age": 30,
  "married": True,
  "divorced": False,
  "children": ("Ann","Billy"),
  "pets": None,
  "cars": [
    {"model": "BMW 230", "mpg": 27.5},
    {"model": "Ford Edge", "mpg": 24.1}
  ]
}

print(json.dumps(x))

{"name": "John", "age": 30, "married": true, "divorced": false, "children": ["Ann", "Billy"], "pets": null, "cars": [{"model": "BMW 230", "mpg": 27.5}, {"model": "Ford Edge", "mpg": 24.1}]}


## Format the Result
The example above prints a JSON string, but it is not very easy to read, with no indentations and line breaks.

The json.dumps() method has parameters to make it easier to read the result:

In [52]:
json.dumps(x, indent=4)

'{\n    "name": "John",\n    "age": 30,\n    "married": true,\n    "divorced": false,\n    "children": [\n        "Ann",\n        "Billy"\n    ],\n    "pets": null,\n    "cars": [\n        {\n            "model": "BMW 230",\n            "mpg": 27.5\n        },\n        {\n            "model": "Ford Edge",\n            "mpg": 24.1\n        }\n    ]\n}'

In [55]:
# You can also define the separators, default value is (", ", ": "), 
# which means using a comma and a space to separate each object,
# and a colon and a space to separate keys from values:

json.dumps(x, indent=4, separators=(". ", " = "))


'{\n    "name" = "John". \n    "age" = 30. \n    "married" = true. \n    "divorced" = false. \n    "children" = [\n        "Ann". \n        "Billy"\n    ]. \n    "pets" = null. \n    "cars" = [\n        {\n            "model" = "BMW 230". \n            "mpg" = 27.5\n        }. \n        {\n            "model" = "Ford Edge". \n            "mpg" = 24.1\n        }\n    ]\n}'

## Order the Result
The json.dumps() method has parameters to order the keys in the result:

In [57]:
json.dumps(x, indent=4, sort_keys=True)


'{\n    "age": 30,\n    "cars": [\n        {\n            "model": "BMW 230",\n            "mpg": 27.5\n        },\n        {\n            "model": "Ford Edge",\n            "mpg": 24.1\n        }\n    ],\n    "children": [\n        "Ann",\n        "Billy"\n    ],\n    "divorced": false,\n    "married": true,\n    "name": "John",\n    "pets": null\n}'

# Try Except

The try block lets you test a block of code for errors.

The except block lets you handle the error.

The finally block lets you execute code, regardless of the result of the try- and except blocks.

Exception Handling
When an error occurs, or exception as we call it, Python will normally stop and generate an error message.

These exceptions can be handled using the try statement:

In [59]:
try:
  print(p)
except:
  print("An exception occurred")

An exception occurred


## Many Exceptions
You can define as many exception blocks as you want, e.g. if you want to execute a special block of code for a special kind of error:

In [62]:
try:
  print(Q)
except NameError:
  print("Variable Q is not defined")
except:
  print("Something else went wrong")

Variable Q is not defined


## Else
You can use the else keyword to define a block of code to be executed if no errors were raised:

In [64]:
try:
  print("Hello")
except:
  print("Something went wrong")
else:
  print("Nothing went wrong")

Hello
Nothing went wrong


# User Input

Python allows for user input.

That means we are able to ask the user for input.

In [65]:
username = input("Enter username:")
print("Username is: " + username)

Enter username:ali
Username is: ali


# String Formatting

To make sure a string will display as expected, we can format the result with the format() method.

String format()
The format() method allows you to format selected parts of a string.

Sometimes there are parts of a text that you do not control, maybe they come from a database, or user input?

To control such values, add placeholders (curly brackets {}) in the text, and run the values through the format() method:

In [69]:
price = 49
txt = "The price is {} dollars"
print(txt.format(price))
txt = "The price is {:.2f} dollars"
print(txt.format(price))

The price is 49 dollars
The price is 49.00 dollars


## Multiple Values
If you want to use more values, just add more values to the format() method:

In [70]:
quantity = 3
itemno = 567
price = 49
myorder = "I want {} pieces of item number {} for {:.2f} dollars."
print(myorder.format(quantity, itemno, price))

I want 3 pieces of item number 567 for 49.00 dollars.


## Index Numbers
You can use index numbers (a number inside the curly brackets {0}) to be sure the values are placed in the correct placeholders:

Example


## Named Indexes
You can also use named indexes by entering a name inside the curly brackets {carname}, but then you must use names when you pass the parameter values txt.format(carname = "Ford"):

In [74]:
quantity = 3
itemno = 567
price = 49
myorder = "I want {0} pieces of item number {1} for {2:.2f} dollars."
print(myorder.format(quantity, itemno, price))

age = 36
name = "John"
txt = "His name is {1}. {1} is {0} years old."
print(txt.format(age, name))


myorder = "I have a {carname}, it is a {model}."
print(myorder.format(carname = "Ford", model = "Mustang"))

I want 3 pieces of item number 567 for 49.00 dollars.
His name is John. John is 36 years old.
I have a Ford, it is a Mustang.


# File Open

## File Handling
The key function for working with files in Python is the open() function.

The open() function takes two parameters; filename, and mode.

There are four different methods (modes) for opening a file:

"r" - Read - Default value. Opens a file for reading, error if the file does not exist

"a" - Append - Opens a file for appending, creates the file if it does not exist

"w" - Write - Opens a file for writing, creates the file if it does not exist

"x" - Create - Creates the specified file, returns an error if the file exists

In addition you can specify if the file should be handled as binary or text mode

"t" - Text - Default value. Text mode

"b" - Binary - Binary mode (e.g. images)

In [78]:
f = open("demofile.txt", "rt") 
# Because "r" for read, and "t" for text are the default values,
# you do not need to specify them.



## Open a File on the Server
To open the file, use the built-in open() function.

The open() function returns a file object, which has a read() method for reading the content of the file:

In [80]:
f = open("demofile.txt", "r")
print(f.read())

Hello! Welcome to demofile.txt
This file is for testing purposes.
Good Luck!


## Read Only Parts of the File
By default the read() method returns the whole text, but you can also specify how many characters you want to return:

In [82]:
f = open("demofile.txt", "r")
print(f.read(5))

Hello


## Read Lines
You can return one line by using the readline() method:

## Close Files
It is a good practice to always close the file when you are done with it.

In [85]:
f = open("demofile.txt", "r")
print(f.readline())
print(f.readline())
print(f.readline())

Hello! Welcome to demofile.txt

This file is for testing purposes.

Good Luck!


In [88]:
f = open("demofile.txt", "r")
for x in f:
  print(x)

f.close()


Hello! Welcome to demofile.txt

This file is for testing purposes.

Good Luck!


## Write to an Existing File
To write to an existing file, you must add a parameter to the open() function:

"a" - Append - will append to the end of the file

"w" - Write - will overwrite any existing content

In [89]:
f = open("demofile2.txt", "a")
f.write("Now the file has more content!")
f.close()

#open and read the file after the appending:
f = open("demofile2.txt", "r")
print(f.read())

Now the file has more content!


In [90]:
f = open("demofile3.txt", "w")
f.write("Woops! I have deleted the content!")
f.close()

#open and read the file after the appending:
f = open("demofile3.txt", "r")
print(f.read())

Woops! I have deleted the content!


Note: the "w" method will overwrite the entire file.

## Create a New File
To create a new file in Python, use the open() method, with one of the following parameters:

"x" - Create - will create a file, returns an error if the file exist

"a" - Append - will create a file if the specified file does not exist

"w" - Write - will create a file if the specified file does not exist

In [109]:
f = open("myfile2.txt", "w")
f.close()

## Delete a File
To delete a file, you must import the OS module, and run its os.remove() function:

In [110]:
import os
os.remove("myfile2.txt")

## Check if File exist:
To avoid getting an error, you might want to check if the file exists before you try to delete it:

In [112]:
import os
if os.path.exists("myfile2.txt"):
  os.remove("myfile2.txt")
else:
  print("The file does not exist")

The file does not exist


## Delete Folder
To delete an entire folder, use the os.rmdir() method:

In [2]:
import os
# os.rmdir("myfolder")

# How to Add Two Numbers in Python

Learn how to add two numbers in Python.

Use the + operator to add two numbers:

In [3]:
x = 5
y = 10
print(x + y)

15


## Add Two Numbers with User Input

In this example, the user must input two numbers. Then we print the sum by calculating (adding) the two numbers:

Example


In [4]:
x = input("Type a number: ")
y = input("Type another number: ")

sum = int(x) + int(y)

print("The sum is: ", sum)

Type a number: 6
Type another number: 7
The sum is:  13


# How to Remove Duplicates From a Python List


In [5]:
mylist = ["a", "b", "a", "c", "c"]
mylist = list(dict.fromkeys(mylist))
print(mylist)

['a', 'b', 'c']


In [7]:
def my_function(x):
  return list(dict.fromkeys(x))

mylist = my_function(["a", "b", "a", "c", "c"])

print(mylist)

['a', 'b', 'c']


## Reverse a String in Python

There is no built-in function to reverse a String in Python.

The fastest (and easiest?) way is to use a slice that steps backwards, -1.

In [9]:
txt = "Hello World"[::-1]
print(txt)

dlroW olleH


In [10]:
txt = "Hello World"
print(txt[::-1])

dlroW olleH
